courses/intro-to-spark/book/05_Proof_Of_Functional_Correctness.rst

```diff
@@ -969,10 +969,50 @@ condition` stating what part of the array has not been modified so far:
 .. code-block:: ada
 
              pragma Loop_Invariant
-               (for all J in K + 1 .. A'Last => A (J) = A'Loop_Entry (J));
+               (for all J in K .. A'Last => A (J) = (if J > K then A'Loop_Entry (J)));
 
 The user-provided and the internally-generated loop invariants are then used to
-prove ``PRESERVE``.
+prove ``PRESERVE``. In more complex cases, the heuristics used by GNATprove to
+generate the frame condition are not sufficient, and a user must provide one as
+loop invariant. For example, consider a version of ``Map`` where the result of
+applying ``F`` to an element at index ``K`` is stored at index ``K-1``:
+
+.. code:: ada spark-report-all
+
+    package Show_Map is
+
+       type Nat_Array is array (Positive range <>) of Natural;
+
+       function F (V : Natural) return Natural is
+         (if V /= Natural'Last then V + 1 else V);
+
+       procedure Map (A : in out Nat_Array);
+
+    end Show_Map;
+
+    package body Show_Map is
+
+       procedure Map (A : in out Nat_Array) is
+          A_I : constant Nat_Array := A with Ghost;
+       begin
+          for K in A'Range loop
+             if K /= A'First then
```

Comment:
I'd rather adapt the range `for K in` to exclude the first element instead of using this `if` inside a loop. However, I'm maybe missing the point of this code example.

Reply:
you can't without modifying elsewhere the example, as you'll need to write "K+1
" which by itself may overflow if K = Integer'Last. :-)

You can see that if you try in the live version.

Reply:
You mean, modifying to something like this?

```
      if A'First /= Positive'Last then
        for K in A'First + 1 .. A'Last loop
```

Reply:
yes, that would work, but I don't find it more readable, do you?

Reply:
I agree, readability is the same. However, that modification makes the code more efficient. Since I care about efficiency, seeing an avoidable `if` inside a `loop` sends shivers down my spine. ;-)

Anyway, no need to change the code if you don't want to.

Reply:
I'm sure the compiler optimizations will figure it out. :-)
Feel free to modify it, at this stage I'm inclined to leave it as it is.